driver/src/orderbook/streamed/state.rs

```diff
@@ -0,0 +1,561 @@
+        self.0
+            .iter()
+            .find(|(id_, _)| *id_ == id)
+            .map(|(_, address)| *address)
+    }
+
+    fn id(&self, address: TokenAddress) -> Option<TokenId> {
+        self.0
+            .iter()
+            .find(|(_, address_)| *address_ == address)
+            .map(|(token_id, _)| *token_id)
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+struct Order {
+    buy_token: TokenId,
+    sell_token: TokenId,
+    valid_from: BatchId,
+    valid_until: BatchId,
+    price_numerator: u128,
+    price_denominator: u128,
+    used_amount: u128,
+}
+
+impl Order {
+    fn has_limited_amount(&self) -> bool {
+        self.price_numerator != std::u128::MAX && self.price_denominator != std::u128::MAX
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+struct Flux {
+    amount: U256,
+    batch_id: BatchId,
+}
+
+#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize)]
+struct Balance {
+    balance: U256,
+    pending_deposit: Option<Flux>,
+    pending_withdraw: Option<Flux>,
+}
+
+impl Balance {
+    fn deposit(&mut self, new_deposit: Flux, current_batch_id: BatchId) {
+        self.update_deposit_balance(current_batch_id);
+        match self.pending_deposit.as_mut() {
+            Some(deposit) => {
+                deposit.amount += new_deposit.amount;
+                deposit.batch_id = new_deposit.batch_id
+            }
+            None => self.pending_deposit = Some(new_deposit),
+        }
+    }
+    fn update_deposit_balance(&mut self, current_batch_id: BatchId) {
+        match self.pending_deposit {
+            Some(ref deposit) if deposit.batch_id < current_batch_id => {
+                self.balance += deposit.amount;
+                self.pending_deposit = None;
+            }
+            _ => (),
+        };
+    }
+    fn current_balance(&self, current_batch_id: BatchId) -> U256 {
+        let mut balance = self.balance;
+        if let Some(ref flux) = self.pending_deposit {
+            if flux.batch_id < current_batch_id {
+                balance = balance.saturating_add(flux.amount);
+            }
+        }
+        if let Some(ref flux) = self.pending_withdraw {
+            if flux.batch_id < current_batch_id {
+                balance = balance.saturating_sub(flux.amount);
+            }
+        }
+        balance
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::models::AccountState;
+
+    fn address(n: u64) -> Address {
+        Address::from_low_u64_be(n)
+    }
+
+    fn state_with_fee() -> State {
+        let mut state = State::default();
+        let event = TokenListing {
+            token: address(0),
+            id: 0,
+        };
+        state.event(&Event::TokenListing(event), 0).unwrap();
+        state
+    }
+
+    fn account_state(state: &State, batch_id: BatchId) -> AccountState {
+        AccountState(state.account_state(batch_id).collect())
+    }
```

Comment:
There is no practical difference. To me this just felt more natural and it is slightly less to type because it doesn't start a new `impl` block.
Even for non test `new` you could implement it as a free standing function but then it would be more awkward with imports but in the test there is no importing.

Reply:
Here I would argue `state.account_state` should return an `AccountState`.